metrics-api/src/main/java/org/killbill/commons/metrics/api/MetricRegistry.java

```diff
@@ -38,52 +38,31 @@ public interface MetricRegistry {
      * @param supplier the underlying Gauge
      * @return a new or pre-existing {@link Gauge}
      */
-    <T> Gauge<T> gauge(String name, Gauge<T> supplier);
+    Gauge gauge(String name, String ...labelNames);
+
+    Gauge gauge(String name, DoubleSupplier supplier, String ...labelNames);
 
     /**
      * Return the {@link Histogram} registered under this name; or create and register
-     * a new {@link Histogram} if none is registered.
+     * a new {@link Histogram} if none is registered using the implementation's default
+     * bucket distribution.
      *
      * @param name the name of the metric
      * @return a new or pre-existing {@link Histogram}
      */
-    Histogram histogram(String name);
+    Histogram histogram(String name, String ...labelNames);
 
-    /**
-     * Return the {@link Meter} registered under this name; or create and register
-     * a new {@link Meter} if none is registered.
-     *
-     * @param name the name of the metric
-     * @return a new or pre-existing {@link Meter}
-     */
-    public Meter meter(String name);
+    Histogram histogram(String name, double[] buckets, String ...labelNames);
 
-    /**
-     * Return the {@link Timer} registered under this name; or create and register
-     * a new {@link Timer} if none is registered.
-     *
-     * @param name the name of the metric
-     * @return a new or pre-existing {@link Timer}
-     */
-    Timer timer(String name);
+    Summary summary(String name, String ...labelNames);
+
+    Summary summary(String name, double[] quantiles, String ...labelNames);
 
     /**
      * Removes the metric with the given name.
      *
      * @param name the name of the metric
      * @return whether or not the metric was removed
      */
-    boolean remove(String name);
-
-    Map<String, ?> getMetrics();
```

Comment:
NACK -- we need to be able to retrieve the metrics (e.g. for `MetricRegistrySerializer` or for `org.killbill.billing.server.metrics.KillbillPluginsMetricRegistry`).

Reply:
I was trying to simplify the interface as much as possible. Seems like for `KillbillPluginsMetricRegistry`, `getMetrics` it's just a pass through, I couldn't really find where the methods is being used directly.

I also couldn't find a definition for the serializer. How's that used?

Reply:
`KillbillPluginsMetricRegistry` is a wrapper around `OSGISingleServiceRegistration<MetricRegistry>` (`MetricRegistryServiceRegistration`), which takes care of discovering the services exported by the bundles (currently at most 1, i.e. the `KillBillCodahaleMetricRegistry` one, exported by the `metrics` bundle, although as discussed, we night change that).

This is used in [MetricRegistrySerializer](https://github.com/killbill/killbill-commons/blob/903411dc937c9d91e3f0a9d6bfc7eb201c062fc5/metrics/src/main/java/org/killbill/commons/metrics/servlets/MetricsJacksonModule.java#L245) for instance for the `MetricsServlet` (`/1.0/metrics` endpoint).

Reply:
The serializer you pointed is tightly coupled to Dropwizard Metrics. That wouldn't work for something else that doesn't produce all those rates and quantiles metrics.

Reply:
We need to find a way to make these changes _somewhat_ backward compatible though, we wouldn't be able to drop that endpoint entirely.

Reply:
I was looking into the implementation of the serializer module as a whole, and a lot of what's used wouldn't be available anymore, not just the `getMetrics`, but also the [rates](https://github.com/killbill/killbill-commons/blob/903411dc937c9d91e3f0a9d6bfc7eb201c062fc5/metrics/src/main/java/org/killbill/commons/metrics/servlets/MetricsJacksonModule.java#L176-L181), [mean, min, max, snapshot](https://github.com/killbill/killbill-commons/blob/903411dc937c9d91e3f0a9d6bfc7eb201c062fc5/metrics/src/main/java/org/killbill/commons/metrics/servlets/MetricsJacksonModule.java#L138-L141) and company, which are all Dropwizard specifc.

A simple option would be to add `getSerializer` method to MetricRegistry. That Jackson module then can be easily changed to use that serializer. But that would introduce a dependency to jackson in the metrics piece, most likely.

Another option is having the implementation bundles exposing a Servlet that would be mounted?

Thoughts?

Reply:
Ignoring this servlet for a moment, we still need a way to _get_ the metrics out of the registry though, for the _sinks_ (_reporters_)?

e.g. 
https://github.com/killbill/killbill-platform/blob/786cca00de17da3af74c0d8872268644c8d353a0/osgi-bundles/bundles/prometheus/src/main/java/org/killbill/billing/osgi/bundles/prometheus/KillBillCollector.java#L55-L69

Otherwise, the bundle implementing the registry must take care of the reporting as well, but this seems like an unnecessary restriction. Also, I'm not sure how we would implement metrics-based healthchecks, like https://github.com/killbill/killbill-commons/blob/903411dc937c9d91e3f0a9d6bfc7eb201c062fc5/jdbi/src/main/java/org/killbill/commons/jdbi/hikari/KillBillHealthChecker.java#L106



Also, looking at other similar projects:
 https://github.com/Netflix/servo/blob/98d4bccbbd4699128c8ee60e67aab1a14b0c410e/servo-core/src/main/java/com/netflix/servo/MonitorRegistry.java#L30

Reply:
> Otherwise, the bundle implementing the registry must take care of the reporting as well

I think that should be the case. Otherwise, we'll have to expose all possible details about a metric in the kb metrics api interface. The problem being that those details are different, depending on the implementation used for metrics.

As an example, Dropwizard Meters (which are equivalent to regular counters in other libraries), keep a lot more info in memory than other implementations (like rates). If we have separate sinks for a dropwizard Meter, than sending that info to a storage might make sense, but for prometheus, not so.

I think the issue in general, is that the current interfaces, and allowing separate sink implementations, make it very hard to have an implementation that's not based in Dropwizard, like prometheus.

We should talk more about this in person to see where we go from here.